ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/plugins/api/ApplicationPluginBuilder.kt

```diff
@@ -82,151 +50,175 @@ public abstract class PluginBuilder<PluginConfig : Any> internal constructor(
+        )
     }
 
     /**
      * Executes specific actions after all [targetPlugins] are executed.
      *
-     * @param targetPlugins Plugins that need to be executed before your current [PluginBuilder].
+     * @param targetPlugins Plugins that need to be executed before your current [ApplicationPluginBuilder].
      * @param build Defines the code of your plugin that needs to be executed after [targetPlugins].
      *
      * Note: you can define multiple actions inside a [build] callback for multiple stages of handling an HTTP call
      * (such as [onCall], [onCallRespond], and so on). These actions are executed right after all actions defined
      * by the given [plugin] are already executed in the same stage.
      **/
-    public fun afterPlugins(
+    public fun after(
         vararg targetPlugins: Plugin<*, *, PluginInstance>,
-        build: AfterPluginBuilder.() -> Unit
+        build: AfterPluginsBuilder<PluginConfig>.() -> Unit
     ) {
         pipelineHandlers.add { pipeline ->
-            AfterPluginBuilder(this, targetPlugins.map { pipeline.plugin(it).builder }).build()
+            AfterPluginsBuilder(this, targetPlugins.map { pipeline.plugin(it).builder }).build()
         }
     }
 
     /**
      * Executes specific actions before all [targetPlugins] are executed.
      *
-     * @param targetPlugins Plugins that need to be executed after your current [PluginBuilder].
+     * @param targetPlugins Plugins that need to be executed after your current [ApplicationPluginBuilder].
      * @param build Defines the code of your plugin that needs to be executed before [targetPlugins].
      *
      * Note: you can define multiple actions inside a [build] callback for multiple stages of handling an HTTP call
      * (such as [onCall], [onCallRespond], and so on) and each of these actions will be executed right before all actions defined
      * by the given [targetPlugins] were already executed in the same stage.
      **/
-    public fun beforePlugins(
+    public fun before(
         vararg targetPlugins: Plugin<*, *, PluginInstance>,
-        build: BeforePluginsBuilder.() -> Unit
+        build: BeforePluginsBuilder<PluginConfig>.() -> Unit
     ) {
         pipelineHandlers.add { pipeline ->
             BeforePluginsBuilder(this, targetPlugins.map { pipeline.plugin(it).builder }).build()
         }
     }
 
+    public fun <EventConfig> on(
+        event: Event<EventConfig>,
+        config: EventConfig
+    ) {
+        events.add(EventHandler(event, config))
+    }
+
     override fun applicationShutdownHook(hook: (Application) -> Unit) {
         environment?.monitor?.subscribe(ApplicationStopped) { app ->
             hook(app)
         }
     }
+
+    /**
+     * Specifies how to modify sending data within an HTTP call for the current [ApplicationPluginBuilder].
+     * @see OnCallRespond
+     **/
+    public override val onCallRespond: OnCallRespond<PluginConfig> = object : OnCallRespond<PluginConfig> {
```

Comment:
Just a question: what would happen if I type `onCallResp<caret>` and call the completion, and rapidly press Enter?
Would it become `onCallRespond { call, data -> ... }` (the method) or just `onCallRespond` (the object) ?

Reply:
Note: we can think of showing just 2 options in the plugin :
a ) `onCallRespond { call, data -> ... }`
b) `onCallRespond.afterTransform { call, data -> ... }` , because the only purpose of the `onCallRespond` object is calling `.afterTransform` method, right :)

Reply:
Please feel free to create a ticket on me to add a completion hook in the plugin once we stick with this implementation.

Reply:
> can order be tweaked?

Yes. Please create a ticket and ping me once this API is finalized